File: robocode-tankroyale-bot-api-java/src/main/java/dev/robocode/tankroyale/botapi/internal/EventQueue.java
package dev.robocode.tankroyale.botapi.internal;

import dev.robocode.tankroyale.botapi.IBaseBot;
import dev.robocode.tankroyale.botapi.events.*;

import java.util.ArrayList;
import java.util.List;
import java.util.SortedMap;
import java.util.concurrent.ConcurrentSkipListMap;
import java.util.concurrent.CopyOnWriteArrayList;

final class EventQueue {

  private final BaseBotInternals baseBotInternals;
  private final BotEventHandlers botEventHandlers;

  private final SortedMap<Integer, List<BotEvent>> eventMap = new ConcurrentSkipListMap<>();

  public EventQueue(BaseBotInternals baseBotInternals, BotEventHandlers botEventHandlers) {
    this.baseBotInternals = baseBotInternals;
    this.botEventHandlers = botEventHandlers;
  }

  void clear() {
    eventMap.clear();
    baseBotInternals.getConditions().clear(); // conditions are added in the bot's run() method each round
  }

  void addEventsFromTick(TickEvent event, IBaseBot baseBot) {
    addEvent(event, baseBot);
    event.getEvents().forEach(evt -> addEvent(evt, baseBot));

    addCustomEvents(baseBot);
  }

  void dispatchEvents() {
    // Publish all event in the order of the keys, i.e. event priority order
    for (List<BotEvent> events : eventMap.values()) {
      ArrayList<BotEvent> eventsCopy = new ArrayList<>(events);
      events.clear();

      for (BotEvent event : eventsCopy) {
        try {
          botEventHandlers.fire(event);
        } catch (RescanException ignore) {}
      }
    }
  }

  private void addEvent(BotEvent event, IBaseBot baseBot) {
    int priority = getPriority(event, baseBot);
    List<BotEvent> botEvents = eventMap.get(getPriority(event, baseBot));
    if (botEvents == null) {
      botEvents = new CopyOnWriteArrayList<>();
      eventMap.put(priority, botEvents);
    }
    botEvents.add(event);
  }

  private void addCustomEvents(IBaseBot baseBot) {
    baseBotInternals
        .getConditions()
        .forEach(
            condition -> {
              if (condition.test()) {
                addEvent(
                    new CustomEvent(baseBotInternals.getCurrentTick().getTurnNumber(), condition),
                    baseBot);
              }
            });
  }

  private static int getPriority(BotEvent event, IBaseBot baseBot) {
    if (event instanceof TickEvent) {
      return EventPriority.onTick;
    } else if (event instanceof ScannedBotEvent) {
      return EventPriority.onScannedBot;
    } else if (event instanceof HitBotEvent) {
      return EventPriority.onHitBot;
    } else if (event instanceof HitWallEvent) {
      return EventPriority.onHitWall;
    } else if (event instanceof BulletFiredEvent) {
      return EventPriority.onBulletFired;
    } else if (event instanceof BulletHitWallEvent) {
      return EventPriority.onBulletHitWall;
    } else if (event instanceof BulletHitBotEvent) {
      BulletHitBotEvent bulletEvent = (BulletHitBotEvent) event;
      if (bulletEvent.getVictimId() == baseBot.getMyId()) {
        return EventPriority.onHitByBullet;
      } else {
        return EventPriority.onBulletHit;
      }
    } else if (event instanceof BulletHitBulletEvent) {
      return EventPriority.onBulletHitBullet;
    } else if (event instanceof DeathEvent) {
      DeathEvent deathEvent = (DeathEvent) event;
      if (deathEvent.getVictimId() == baseBot.getMyId()) {
        return EventPriority.onDeath;
      } else {
        return EventPriority.onBotDeath;
      }
    } else if (event instanceof SkippedTurnEvent) {
      return EventPriority.onSkippedTurn;
    } else if (event instanceof CustomEvent) {
      return EventPriority.onCondition;
    } else if (event instanceof WonRoundEvent) {
      return EventPriority.onWonRound;
    } else {
      throw new IllegalStateException("Unhandled event type: " + event);
    }
  }
}

File: robocode-tankroyale-bot-api-dotnet/src/internal/EventQueue.cs
using System;
using System.Collections;
using System.Collections.Concurrent;
using System.Collections.Generic;
using Robocode.TankRoyale.BotApi.Events;

namespace Robocode.TankRoyale.BotApi.Internal
{
  internal sealed class EventQueue
  {
    private const int MaxQueueSize = 256;
    private const int MaxEventAge = 2;

    private readonly BaseBotInternals baseBotInternals;
    private readonly BotEventHandlers botEventHandlers;

    private readonly IDictionary<int, ArrayList> eventsDict = new ConcurrentDictionary<int, ArrayList>();

    private BotEvent currentEvent;

    internal EventQueue(BaseBotInternals baseBotInternals, BotEventHandlers botEventHandlers)
    {
      this.baseBotInternals = baseBotInternals;
      this.botEventHandlers = botEventHandlers;
    }

    public void Clear()
    {
      eventsDict.Clear();
      baseBotInternals.Conditions.Clear(); // conditions might be added in the bot's Run() method each round
      currentEvent = null;
    }

    internal void AddEventsFromTick(TickEvent tickEvent, IBaseBot baseBot)
    {
      AddEvent(tickEvent, baseBot);
      foreach (var botEvent in tickEvent.Events)
      {
        AddEvent(botEvent, baseBot);
      }
      AddCustomEvents(baseBot);
    }

    internal void DispatchEvents(int currentTurn)
    {
      RemoveOldEvents(currentTurn);

      // Handle events in the order of the keys, i.e. event priority order
      var sortedDict = new SortedDictionary<int, ArrayList>(eventsDict);

      foreach (var events in sortedDict.Values)
      {
        for (int i = events.Count - 1; i >= 0; i--)
        {
          try
          {
            var evt = (BotEvent)events[i];

            // Exit if we are inside an event handler handling the current event being fired
            if (currentEvent != null && events.GetType().Equals(currentEvent.GetType()))
              return;

            try
            {
              currentEvent = evt;
              events.RemoveAt(i);
              botEventHandlers.Fire(evt);
            }
            catch (RescanException)
            {
              currentEvent = null;
            }
            catch (Exception)
            {
              currentEvent = null;
              throw;
            }
          }
          catch (System.ArgumentOutOfRangeException)
          {
            continue;
          }
        }
      }
    }

    private void RemoveOldEvents(int currentTurn)
    {
      foreach (var events in eventsDict.Values)
      {
        for (int i = events.Count - 1; i >= 0; i--)
        {
          try
          {
            var evt = (BotEvent)events[i];
            if (!evt.IsCritical && IsOldEvent(evt, currentTurn))
              events.RemoveAt(i);
          }
          catch (System.ArgumentOutOfRangeException)
          {
            continue;
          }
        }
      }
    }

    private static bool IsOldEvent(BotEvent botEvent, int currentTurn)
    {
      return botEvent.TurnNumber + MaxEventAge < currentTurn;
    }

    private void AddEvent(BotEvent botEvent, IBaseBot baseBot)
    {
      if (CountEvents() > MaxQueueSize)
      {
        Console.Error.WriteLine($"Maximum event queue size has been reached: {MaxQueueSize}");
      }
      else
      {
        int priority = GetPriority(botEvent, baseBot);

        ArrayList events;
        eventsDict.TryGetValue(priority, out events);
        if (events == null)
        {
          events = ArrayList.Synchronized(new ArrayList());
          eventsDict.Add(priority, events);
        }
        events.Add(botEvent);
      }
    }

    private int CountEvents()
    {
      int count = 0;
      foreach (var events in eventsDict.Values)
      {
        count += events.Count;
      }
      return count;
    }

    private void AddCustomEvents(IBaseBot baseBot)
    {
      foreach (Events.Condition condition in baseBotInternals.Conditions)
      {
        if (condition.Test())
        {
          AddEvent(new CustomEvent(baseBotInternals.CurrentTick.TurnNumber, condition), baseBot);
        }
      }
    }

    private static int GetPriority(BotEvent botEvent, IBaseBot baseBot)
    {
      switch (botEvent)
      {
        case TickEvent tickEvent:
          return EventPriority.OnTick;
        case ScannedBotEvent scannedBotEvent:
          return EventPriority.OnScannedBot;
        case HitBotEvent hitBotEvent:
          return EventPriority.OnHitBot;
        case HitWallEvent hitWallEvent:
          return EventPriority.OnHitWall;
        case BulletFiredEvent bulletFiredEvent:
          return EventPriority.OnBulletFired;
        case BulletHitWallEvent bulletHitWallEvent:
          return EventPriority.OnBulletHitWall;
        case BulletHitBotEvent bulletHitBotEvent:
          if (bulletHitBotEvent.VictimId == baseBot.MyId)
            return EventPriority.OnHitByBullet;
          else
            return EventPriority.OnBulletHit;
        case BulletHitBulletEvent bulletHitBulletEvent:
          return EventPriority.OnBulletHitBullet;
        case DeathEvent deathEvent:
          if (deathEvent.VictimId == baseBot.MyId)
            return EventPriority.OnDeath;
          else
            return EventPriority.OnBotDeath;
        case SkippedTurnEvent skippedTurnEvent:
          return EventPriority.OnSkippedTurn;
        case CustomEvent customEvent:
          return EventPriority.OnCondition;
        case WonRoundEvent wonRoundEvent:
          return EventPriority.OnWonRound;
        default:
          throw new InvalidOperationException("Unhandled event type: " + botEvent);
      }
    }
  }
}